File: src/bot/commands/link/verify.ts
import { MessageFlags, SlashCommandBuilder, type ChatInputCommandInteraction } from 'discord.js';
import { Command } from '../../struct/Command.js';
import { EMOJIS } from '../../../util/emojis.js';
import { isValidTag, normalizeTag } from '../../../util/helper.js';

/**
 * Prove ownership of an account using the in-game API token
 * (Settings → More Settings → API Token). A verified link is trusted for role assignment.
 */
export default class VerifyCommand extends Command {
  public constructor() {
    super('verify', { category: 'link', description: 'Verify ownership of a linked account.' });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('verify')
      .setDescription('Verify ownership of a Clash of Clans account.')
      .addStringOption((opt) => opt.setName('tag').setDescription('Player tag.').setRequired(true))
      .addStringOption((opt) =>
        opt.setName('token').setDescription('In-game API token (Settings → More Settings → API Token).').setRequired(true)
      );
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    if (!isValidTag(args.tag)) {
      return interaction.reply({ content: 'That is not a valid player tag.', flags: MessageFlags.Ephemeral });
    }
    await interaction.deferReply({ flags: MessageFlags.Ephemeral });
    const tag = normalizeTag(args.tag);

    const ok = await this.client.coc.verifyPlayerToken(tag, String(args.token).trim());
    if (!ok) return interaction.editReply(`${EMOJIS.CROSS} Verification failed. Double-check the tag and API token.`);

    const player = await this.client.coc.getPlayer(tag);
    const count = await this.client.db.playerLinks.countDocuments({ userId: interaction.user.id });
    await this.client.db.playerLinks.updateOne(
      { userId: interaction.user.id, tag },
      {
        $set: { verified: true, name: player?.name ?? tag },
        $setOnInsert: { userId: interaction.user.id, tag, order: count, createdAt: new Date() }
      },
      { upsert: true }
    );

    return interaction.editReply(`${EMOJIS.CHECK} Verified ownership of **${player?.name ?? tag}** (\`${tag}\`).`);
  }
}

File: cogs/link/commands/verify.py
"""/link verify, prove account ownership with the in-game API token."""

import discord

from database.db import get_pool
from utils.embeds import error_embed, success_embed
from utils.tags import is_valid_tag, normalize_tag


async def verify(interaction: discord.Interaction, tag: str, token: str):
    await interaction.response.defer(ephemeral=True)

    if not is_valid_tag(tag):
        await interaction.followup.send(embed=error_embed("That is not a valid player tag."))
        return
    tag = normalize_tag(tag)

    try:
        ok = await interaction.client.coc_client.verify_player_token(tag, token.strip())
    except Exception:
        ok = False
    if not ok:
        await interaction.followup.send(
            embed=error_embed("Verification failed. Double-check the tag and the in-game API token.")
        )
        return

    try:
        player = await interaction.client.coc_client.get_player(tag)
        name = player.name
    except Exception:
        name = tag

    pool = await get_pool()
    # Upsert: verify an existing link, or create a verified one.
    await pool.execute(
        """INSERT INTO linked_accounts (discord_id, tag, name, verified)
           VALUES ($1, $2, $3, TRUE)
           ON CONFLICT (tag) DO UPDATE SET verified = TRUE, name = EXCLUDED.name, discord_id = $1""",
        interaction.user.id,
        tag,
        name,
    )
    await interaction.followup.send(embed=success_embed(f"Verified ownership of **{name}** (`{tag}`)."))
